src/rust/engine/fs/src/store.rs

```diff
@@ -2516,6 +2567,130 @@ mod tests {
     );
   }
 
+  #[test]
+  fn uploads_files_recursively_when_under_three_digests_ignoring_items_already_in_cas() {
+    let dir = TempDir::new("store").unwrap();
+    let cas = StubCAS::empty();
+
+    new_local_store(dir.path())
+      .record_directory(&directory(), false)
+      .wait()
+      .expect("Error storing directory locally");
+    new_local_store(dir.path())
+      .store_file_bytes(str_bytes(), false)
+      .wait()
+      .expect("Error storing file locally");
+
+    new_store(dir.path(), cas.address())
+      .ensure_remote_has_recursive(vec![digest()])
+      .wait()
+      .expect("Error uploading file");
+
+    assert_eq!(cas.write_message_sizes.lock().unwrap().len(), 1);
+    assert_eq!(
```

Comment:
I'd be tempted to extract a method so you can call `assert_contains_exactly(cas, vec![fingerprint()])`, which reads a bit easier than the several `assert_eq!` calls on both size and contents

Reply:
How would you feel about holding off on this until changes to the test data merge? Right now we are checking that the digests exist in the cas and also checking that they are the bytes we expect. To do this, we would probably need to pass a hashmap of fingerprint to bytes to the function that does the asserts. With the new test data, this would simplify things... I could definitely implement it with a hashmap if you would like though 😄

Reply:
Happy to hold off :)